File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/BinaryOperationWithIdenticalExpressions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Tests.TestCases
{
    class Foo { }

    class BinaryOperationWithIdenticalExpressions
    {
        public void doZ() { throw new Exception();}
        public void doW() { throw new Exception();}
        public void Test(bool a, bool b)
        {
            if (a == a)
//                   ^ {{Correct one of the identical expressions on both sides of operator '=='.}}
//              ^ Secondary@-1
            {
                doZ();
            }

            if (a == b || (a == /*comment*/ b))
//                        ^^^^^^^^^^^^^^^^^^^^ {{Correct one of the identical expressions on both sides of operator '||'.}}
//              ^^^^^^ Secondary@-1
            {
                doW();
            }

            int j = 5 / 5; //Noncompliant
            // Secondary@-1
            int k = 5 - 5; //Noncompliant
            // Secondary@-1
            int l = 5 * 5;

            l = 5 | 5; // Noncompliant
            // Secondary@-1
            l |= (l); // Noncompliant
            // Secondary@-1

            int i = 1 << 1;
            i = 1 << 0x1;
            i = 2 << 2; // Compliant

            object.Equals(i, i);
//                        ^ {{Correct one of the identical expressions on both sides of Object.Equals.}}
//                           ^ Secondary@-1

            var o = new object();

            o.Equals(o);
//          ^ {{Correct one of the identical expressions on both sides of Object.Equals.}}
//                   ^ Secondary@-1

            (new object()).Equals(new object());
//          ^^^^^^^^^^^^^^ {{Correct one of the identical expressions on both sides of Object.Equals.}}
//                                ^^^^^^^^^^^^ Secondary@-1

            Foo f;
            f.Equals(f); // Noncompliant
            // Secondary@-1

        }
    }
}

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/BinaryOperationWithIdenticalExpressions.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace Tests.TestCases
    Class BinaryOperationWithIdenticalExpressions
        Public Sub doZ()
        End Sub
        Public Sub doW()
        End Sub
        Public Sub Test(a As Boolean, b As Boolean)

            If (a = a) Then ' Noncompliant {{Correct one of the identical expressions on both sides of operator '='.}}
'               ^^^^^
                doZ()
            End If

            If a = b OrElse a = b Then ' Noncompliant {{Correct one of the identical expressions on both sides of operator 'OrElse'.}}
                doW()
            End If

            Dim j = 5 / 5 ' Noncompliant
            j = 5 \ 5 ' Noncompliant
            j = 5 Mod 5 ' Noncompliant
            Dim k = 5 - 5 ' Noncompliant
            Dim l = 5 * 5


            Dim i = 1 << 1

            i -= i ' Noncompliant
            i += i
            i /= i ' Noncompliant
            i \= i ' Noncompliant
        End Sub
    End Class
End Namespace

